src/main/java/org/candlepin/resource/ConsumerResource.java

```diff
@@ -1604,57 +1598,83 @@ else if (toUpdate.getContentAccessMode() != null) {
+                        throw new NotFoundException(i18n.tr(
+                            "Environment with ID \"{0}\" could not be found.", environmentId));
+                    }
                 }
+                else if (environmentName != null) {
+                    environmentId = this.environmentCurator
+                        .getEnvironmentIdByName(toUpdate.getOwnerId(), environmentName);
 
-                validatedEnvId = environmentId;
-            }
-
-        }
-        else if (environmentName != null) {
-            String envId = this.environmentCurator.
-                getEnvironmentIdByName(toUpdate.getOwnerId(), environmentName);
+                    if (environmentId == null) {
+                        throw new NotFoundException(i18n.tr(
+                            "Environment with name \"{0}\" could not be found.", environmentName));
+                    }
+                }
+                else {
+                    throw new NotFoundException(i18n.tr("Environment could not be" +
+                        " found using provided details."));
+                }
 
-            if (envId == null) {
-                throw new NotFoundException(i18n.tr(
-                    "Environment with name \"{0}\" could not be found.", environmentName));
-            }
+                if (validatedEnvIds.contains(environmentId)) {
+                    throw new ConflictException("Environment \"{0}\"" +
+                        " specified more than once." + environmentId);
+                }
 
-            if (toUpdate.getEnvironmentId() == null || !envId.equals(toUpdate.getEnvironmentId())) {
-                validatedEnvId = envId;
+                validatedEnvIds.add(environmentId);
             }
         }
 
-        if (validatedEnvId != null) {
-            toUpdate.setEnvironmentId(validatedEnvId);
-            // reset content access cert
-            Owner owner = ownerCurator.findOwnerById(toUpdate.getOwnerId());
+        if (validatedEnvIds.size() > 0) {
+            toUpdate.setEnvironmentIds(validatedEnvIds);
 
-            if (owner.isUsingSimpleContentAccess()) {
-                toUpdate.setContentAccessCert(null);
-                this.contentAccessManager.removeContentAccessCert(toUpdate);
+            if (existing) {
+                // reset content access cert
+                Owner owner = ownerCurator.findOwnerById(toUpdate.getOwnerId());
+
+                if (owner.isUsingSimpleContentAccess()) {
+                    toUpdate.setContentAccessCert(null);
+                    this.contentAccessManager.removeContentAccessCert(toUpdate);
+                }
+
+                // lazily regenerate certs, so the client can still work
+                poolManager.regenerateCertificatesOf(toUpdate, true);
             }
 
-            // lazily regenerate certs, so the client can still work
-            poolManager.regenerateCertificatesOf(toUpdate, true);
```

Comment:
We still need to do this every time the environment(s) change, right? The certs will have environment-specific details in them, so I can't imagine we can get away from regenerating the certs when this changes.

@nikosmoum

Reply:
Yes we need to, but only in case of a PUT request (which with the current state of the PR will still happen). But it doesn't have to be done during a POST (registration) call, since there's no certificates yet (this is the only path in which this is not run)